presto-main/src/main/java/io/prestosql/operator/RowReferencePageManager.java

```diff
@@ -0,0 +1,457 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.slice.SizeOf;
+import io.prestosql.array.IntBigArray;
+import io.prestosql.spi.Page;
+import io.prestosql.util.LongBigArrayFIFOQueue;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntIterator;
+import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
+import it.unimi.dsi.fastutil.longs.LongComparator;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static java.lang.Math.toIntExact;
+
+/**
+ * Page buffering manager that enables access to individual rows via stable row IDs. This allows computation to be
+ * built against these row IDs, while still enabling bulk memory optimizations such as compaction and lazy loading
+ * behind the scenes. Callers are responsible for explicitly de-referencing any rows that are no longer needed.
+ */
+public class RowReferencePageManager
+{
+    private static final long INSTANCE_SIZE = ClassLayout.parseClass(RowReferencePageManager.class).instanceSize();
+    private static final long PAGE_ACCOUNTING_INSTANCE_SIZE = ClassLayout.parseClass(PageAccounting.class).instanceSize();
+
+    private final IdRegistry<PageAccounting> pages = new IdRegistry<>();
+    private final IdRegistry<LoadCursor> cursors = new IdRegistry<>();
+    private final RowIdBuffer rowIdBuffer = new RowIdBuffer();
+    private final IntList lazyLoadCandidates = new IntList();
+    private final IntHashSet compactionCandidates = new IntHashSet();
+
+    private long pageBytes;
+
+    public LoadCursor add(Page page)
+    {
+        int pageId = pages.allocateId();
+        PageAccounting pageAccounting = new PageAccounting(pageId, page);
+        pages.set(pageId, pageAccounting);
```

Comment:
This wouldn't be for the "replace case", but only "add new", so it would only have the one ID.  I'm thinking of something like `Map.computeIfAbsent`, but for adding new items